README.md
# SQL Exercises

My attempt at some [SQL Exercises](https://en.wikibooks.org/wiki/SQL_Exercises)
using [squeal-postgresql](https://hackage.haskell.org/package/squeal-postgresql).


CHANGELOG.md
# Revision history for sql-exercises

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


exercises/TheWarehouse.hs
{-# LANGUAGE OverloadedLabels #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE DataKinds #-}
module TheWarehouse where

import Shared
import Squeal.PostgreSQL
import Data.Int (Int64)
import Data.Text (pack)

type instance Columns "warehouses" =
  '[ "code" ::: 'Def :=> 'NotNull 'PGint8
   , "location" ::: 'NoDef :=> 'NotNull 'PGtext
   , "capacity" ::: 'NoDef :=> 'NotNull 'PGint8 ]

type instance Constraints "warehouses" =
  '[ "pk_warehouses" ::: 'PrimaryKey '["code"] ]

type instance Columns "boxes" =
  '[ "code" ::: 'Def :=> 'NotNull 'PGtext
   , "contents" ::: 'NoDef :=> 'NotNull 'PGtext
   , "value" ::: 'NoDef :=> 'NotNull 'PGmoney
   , "warehouse" ::: 'NoDef :=> 'NotNull 'PGint8 ]

type instance Constraints "boxes" =
  '[ "pk_boxes" ::: 'PrimaryKey '["code"]
   , "fk_warehouse" ::: 'ForeignKey '["warehouse"] "warehouses" '["code"] ]

type DB = Public '[ SimpleTable "warehouses", SimpleTable "boxes" ]

-- | 1. Select all warehouses.
ex1 :: Query '[] '[] DB '[]
  '[ "code" ::: 'NotNull 'PGint8
   , "location" ::: 'NotNull 'PGtext
   , "capacity" ::: 'NotNull 'PGint8 ]
ex1 = select_ (#code :* #location :* #capacity) (from (table #warehouses))

-- | 2. Select all boxes with a value larger than $150.
ex2 :: Query '[] '[] DB '[]
  '[ "code" ::: 'NotNull 'PGtext
   , "contents" ::: 'NotNull 'PGtext
   , "value" ::: 'NotNull 'PGmoney
   , "warehouse" ::: 'NotNull 'PGint8 ]
ex2 = select_ (#code :* #contents :* #value :* #warehouse)
  (from (table #boxes)
  & where_ (#value .> inline (Money 15000)))

-- | 3. Select all distinct contents in all the boxes.
ex3 :: Query '[] '[] DB '[]
  '[ "code" ::: 'NotNull 'PGtext
   , "contents" ::: 'NotNull 'PGtext
   , "value" ::: 'NotNull 'PGmoney
   , "warehouse" ::: 'NotNull 'PGint8 ]
ex3 = selectDistinct_ (#code :* #contents :* #value :* #warehouse)
  (from (table #boxes)
  & where_ (#value .> inline (Money 15000)))

-- | 4. Select the average value of all the boxes.
ex4 :: Query '[] '[] DB '[]
  '[ "average_value" ::: 'NotNull 'PGmoney ]
ex4 = select_ (money `cast` avg (All $ numeric `cast` #value) `as` #average_value) (from (table #boxes) & groupBy #code)

-- | 5. Select the warehouse code and the average value of the boxes in each
-- warehouse.
ex5 :: Query '[] '[] DB '[]
  '[ "warehouse_code" ::: 'NotNull 'PGint8
   , "average_box_value" ::: 'Null 'PGmoney ]
ex5 = select_ (#w ! #code `as` #warehouse_code
            :* (money `cast` avg (All $ numeric `cast` (#b ! #value))) `as` #average_box_value)
  (from (table (#warehouses `as` #w) & innerJoin (table (#boxes `as` #b)) (#w ! #code .== #b ! #warehouse))
  & groupBy (#w ! #code))

-- | 6. Same as previous exercise, but select only those warehouses where the
-- average value of the boxes is greater than 150.
ex6 :: Query '[] '[] DB '[]
  '[ "warehouse_code" ::: 'NotNull 'PGint8
   , "average_box_value" ::: 'NotNull 'PGmoney ]
ex6 = select_ (#w ! #code `as` #warehouse_code
            :* (money `cast` avg (All $ numeric `cast` (#b ! #value))) `as` #average_box_value)
  (from (table (#warehouses `as` #w) & innerJoin (table (#boxes `as` #b)) (#w ! #code .== #b ! #warehouse))
  & groupBy (#w ! #code)
  & having ((money `cast` avg (All $ numeric `cast` (#b ! #value))) .> inline (Money 15000)))

-- | 7. Select the code of each box, along with the name of the city the box is located in.
ex7 :: Query '[] '[] DB '[]
  '[ "box_code" ::: 'NotNull 'PGtext
   , "city_name" ::: 'NotNull 'PGtext ]
ex7 = select_ (#b ! #code `as` #box_code :* #w ! #location `as` #city_name) 
  (from (table (#boxes `as` #b) 
        & innerJoin (table (#warehouses `as` #w)) (#b ! #warehouse .== #w ! #code)))

-- | 8. Select the warehouse codes, along with the number of boxes in each
-- warehouse. Optionally, take into account that some warehouses are empty
-- (i.e., the box count should show up as zero, instead of omitting the warehouse
-- from the result).
ex8 :: Query '[] '[] DB '[]
  '[ "warehouse_code" ::: 'NotNull 'PGint8
   , "box_count" ::: 'NotNull 'PGint8 ]
ex8 = select_
   (#w ! #code `as` #warehouse_code
 :* count (All $ #b ! #code) `as` #box_count)
   (from (table (#warehouses `as` #w)
         & leftOuterJoin (table (#boxes `as` #b)) (#b ! #warehouse .== #w ! #code))
   & groupBy (#w ! #code))

-- | 9. Select the codes of all warehouses that are saturated (a warehouse is
-- saturated if the number of boxes in it is larger than the warehouse's capacity).
ex9 :: Query '[] '[] DB '[]
  '[ "warehouse_code" ::: 'NotNull 'PGint8 ]
ex9 = select_
   (#w ! #code `as` #warehouse_code)
   (from (table (#warehouses `as` #w)
         & leftOuterJoin (table (#boxes `as` #b)) (#b ! #warehouse .== #w ! #code))
   & groupBy (#w ! #code :* #w ! #capacity)
   & having (count (All $ #b ! #code) .> #w ! #capacity))

-- | 10. Select the codes of all the boxes located in Chicago.
ex10 :: Query '[] '[] DB '[]
  '[ "warehouse_code" ::: 'NotNull 'PGint8 ]
ex10 = select_
  (#w ! #code `as` #warehouse_code)
  (from (table (#warehouses `as` #w))
  & where_ (#w ! #location .== inline "Chicago"))

-- | 11. Create a new warehouse in New York with a capacity for 3 boxes.
ex11 :: Manipulation '[] DB '[] '[]
ex11 = insertInto #warehouses
  (Values_
    (Default `as` #code
  :* Set (inline "New York") `as` #location
  :* Set (inline (3 :: Int64)) `as` #capacity
    )
  )
  OnConflictDoRaise 
  (Returning_ Nil)

-- | 12. Create a new box, with code "H5RT", containing "Papers" with a value
-- of $200, and located in warehouse 2.
ex12 :: Manipulation '[] DB '[] '[]
ex12 = insertInto #boxes
  (Values_
    (Set (inline "H5RT") `as` #code
  :* Set (inline "Papers") `as` #contents
  :* Set (inline (Money 20000)) `as` #value
  :* Set (inline (2 :: Int64)) `as` #warehouse
    )
  )
  OnConflictDoRaise
  (Returning_ Nil)
    
-- | 13. Reduce the value of all boxes by 15%.
ex13 :: Manipulation '[] DB '[] '[]
ex13 = update_ #boxes 
  (Set (money `cast` ((numeric `cast` #value) * 0.85)) `as` #value) true 

-- | 14. Apply a 20% value reduction to boxes with a value larger than the
-- average value of all the boxes.
-- TODO
ex14 :: Manipulation '[] DB '[] '[]
ex14 = with (q `as` #q) _
  where
    q = queryStatement (select_ (avg (All $ #code) `as` #average_box_value) (from (table #boxes) & groupBy Nil))

-- | 15. Remove all boxes with a value lower than $100.
-- TODO


exercises/TheComputerStore.hs
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE OverloadedLabels #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TypeFamilies #-}
module TheComputerStore where

import Data.Int (Int64)
import GHC.TypeLits (Symbol)
import Squeal.PostgreSQL
import Data.Text (pack)
import Shared

type instance Columns "manufacturers" =
  '[ "code" ::: 'Def :=> 'NotNull 'PGint8
   , "name" ::: 'NoDef :=> 'NotNull 'PGtext
   ]

type instance Constraints "manufacturers" =
  '[ "pk_manufacturers" ::: 'PrimaryKey '["code"]
   ]

type instance Columns "products" =
  '[ "code" ::: 'Def :=> 'NotNull 'PGint8
   , "name" ::: 'NoDef :=> 'NotNull 'PGtext
   , "price" ::: 'NoDef :=> 'NotNull 'PGmoney
   , "manufacturer" ::: 'NoDef :=> 'NotNull 'PGint8 
   ]

type instance Constraints "products" =
  '[ "pk_products" ::: 'PrimaryKey '["code"]
   , "fk_manufacturer" ::: 'ForeignKey '["manufacturer"] "manufacturers" '["code"]
   ]

type DB = Public '[ SimpleTable "manufacturers", SimpleTable "products" ]

-- | 1. Select the names of all the products in the store.
ex1 :: Query '[] '[] DB '[]
  '["product_name" ::: 'NotNull 'PGtext]
ex1
  = select_ (#name `as` #product_name) 
    (from (table #products))

-- | 2. Select the names and the prices of all the products in the store.
ex2 :: Query '[] '[] DB '[]
  '["product_name" ::: 'NotNull 'PGtext, "product_price" ::: 'NotNull 'PGmoney]
ex2
  = select_ (#name `as` #product_name
          :* #price `as` #product_price) 
    (from (table #products))

-- | 3. Select the name of the products with a price less than or equal to $200.
ex3 :: Query '[] '[] DB '[] '["product_name" ::: 'NotNull 'PGtext]
ex3
  = select_ (#name `as` #product_name)
    (from (table #products) & where_ (#price .< inline (Money 200)))

-- | 4. Select all the products with a price between $60 and $120.
ex4 :: Query '[] '[] DB '[]
  '[ "product_code" ::: 'NotNull 'PGint8
   , "product_name" ::: 'NotNull 'PGtext
   , "product_price" ::: 'NotNull 'PGmoney
   , "product_manufacturer" ::: 'NotNull 'PGint8
   ]
ex4
  = select_ (#code `as` #product_code
          :* #name `as` #product_name
          :* #price `as` #product_price
          :* #manufacturer `as` #product_manufacturer)
    (from (table #products) & where_ (#price `between` (inline (Money 60), inline (Money 200))))

-- | 5. Select the name and price in cents (i.e., the price must be multiplied by 100).
ex5 :: Query '[] '[] DB '[]
  '[ "product_name" ::: 'NotNull 'PGtext
   , "product_price_in_cents" ::: 'NotNull 'PGint8
   ]
ex5
  = select_ (#name `as` #product_name
          :* ((#price & cast int8) * 2) `as` #product_price_in_cents)
    (from (table #products))

-- | 6. Compute the average price of all the products.
ex6 :: Query '[] '[] DB '[]
  '[ "average_product_price" ::: 'Null 'PGnumeric
   ]
ex6
  = select_ (avg (All (int8 `cast` (#p ! #price))) `as` #average_product_price)
    (from (table (#products `as` #p)) & groupBy Nil)

-- | 7. Compute the average price of all products with manufacturer code equal to 2.
ex7 :: Query '[] '[] DB '[]
  '[ "average_product_price" ::: 'Null 'PGmoney
   ]
ex7
 = select_ (money `cast` avg (All (int8 `cast` (#p ! #price))) `as` #average_product_price)
   (from (table (#products `as` #p)) & where_ (#p ! #manufacturer .== inline (2 :: Int64)) & groupBy Nil)

-- | 8. Compute the number of products with a price larger than or equal to $180.
ex8 :: Query '[] '[] DB '[]
  '[ "product_count" ::: 'NotNull 'PGint8 ]
ex8
  = select_ (countStar `as` #product_count)
    (from (table (#products `as` #p)) & where_ (#p ! #price .>= inline (Money 180)) & groupBy Nil)

-- | 9. Select the name and price of all products with a price larger than or
-- equal to $180, and sort first by price (in descending order), and then by
-- name (in ascending order).
ex9 :: Query '[] '[] DB '[]
  '[ "product_name" ::: 'NotNull 'PGtext
   , "product_price" ::: 'NotNull 'PGmoney
   ]
ex9
  = select_ (#p ! #name `as` #product_name
          :* #p ! #price `as` #product_price)
    (from (table (#products `as` #p)) 
      & where_ (#p ! #price .>= inline (Money 180)) 
      & orderBy [Desc (#p ! #price), Asc (#p ! #name)])

-- | 10. Select all the data from the products, including all the data for
-- each product's manufacturer.
ex10 :: Query '[] '[] DB '[]
  '[ "product_code" ::: 'NotNull 'PGint8
   , "product_name" ::: 'NotNull 'PGtext
   , "product_price" ::: 'NotNull 'PGmoney
   , "product_manufacturer" ::: 'NotNull 'PGint8
   , "product_manufacturer_name" ::: 'NotNull 'PGtext
   ]
ex10
  = select_ (#p ! #code `as` #product_code
          :* #p ! #name `as` #product_name
          :* #p ! #price `as` #product_price
          :* #p ! #manufacturer `as` #product_manufacturer
          :* #m ! #name `as` #product_manufacturer_name)
    (from (table (#products `as` #p)
          & innerJoin (table (#manufacturers `as` #m)) (#m ! #code .== #p ! #manufacturer)))

-- | 11. Select the product name, price, and manufacturer name of all the products.
ex11 :: Query '[] '[] DB '[]
  '[ "product_name" ::: 'NotNull 'PGtext
   , "product_price" ::: 'NotNull 'PGmoney
   , "product_manufacturer_name" ::: 'NotNull 'PGtext
   ]
ex11
  = select_ (#p ! #name `as` #product_name
          :* #p ! #price `as` #product_price
          :* #m ! #name `as` #product_manufacturer_name)
    (from (table (#products `as` #p)
          & innerJoin (table (#manufacturers `as` #m)) (#p ! #manufacturer .== #m ! #code)))

-- | 12. Select the average price of each manufacturer's products, showing only
-- the manufacturer's code.
ex12 :: Query '[] '[] DB '[]
  '[ "manufacturer_code" ::: 'NotNull 'PGint8
   , "average_product_price" ::: Null 'PGmoney ]
ex12
 = select_ (#m ! #code `as` #manufacturer_code
         :* money `cast` avg (All (int8 `cast` (#p ! #price))) `as` #average_product_price)
    (from (table (#manufacturers `as` #m)
                  & leftOuterJoin (table (#products `as` #p)) (#p ! #manufacturer .== #m ! #code))
     & groupBy (#m ! #code))

-- | 13. Select the average price of each manufacturer's products, showing the manufacturer's name.
ex13 :: Query '[] '[] DB '[]
  '[ "manufacturer_name" ::: 'NotNull 'PGtext
   , "average_product_price" ::: 'NotNull 'PGmoney ]
ex13
  = select_ (#m ! #name `as` #manufacturer_name
          :* money `cast` avg (All (int8 `cast` (#p ! #price))) `as` #average_product_price)
    (from (table (#manufacturers `as` #m)
                & leftOuterJoin (table (#products `as` #p)) (#m ! #code .== #p ! #manufacturer))
    & groupBy (#m ! #code :* #m ! #name))

-- | 14. Select the names of manufacturer whose products have an average price
-- larger than or equal to $150.
ex14 :: Query '[] '[] DB '[]
  '[ "manufacturer_name" ::: 'NotNull 'PGtext ]
ex14
  = select_ (#m ! #name `as` #manufacturer_name)
    (from (table (#manufacturers `as` #m)
                & leftOuterJoin (table (#products `as` #p)) (#p ! #manufacturer .== #m ! #code))
     & groupBy (#m ! #code :* #m ! #name)
     & having (money `cast` avg (All (int8 `cast` (#p ! #price))) .>= inline (Money 150) ))

-- | 15. Select the name and price of the cheapest product.
ex15 :: Query '[] '[] DB '[]
  '[ "product_name" ::: 'NotNull 'PGtext
   , "product_price" ::: 'NotNull 'PGmoney ]
ex15
  = select_ (#p ! #name `as` #product_name
          :* #p ! #price `as` #product_price)
    (from (table (#products `as` #p))
    & orderBy [Asc $ #p ! #price]
    & limit 1)

-- | 16. Select the name of each manufacturer along with the name and price of its most expensive product.
ex16 :: Query '[] '[] DB '[]
  '[ "manufacturer_name" ::: 'NotNull 'PGtext
   , "product_price" ::: 'Null 'PGmoney 
   , "product_name" ::: 'Null 'PGtext ]
ex16
  = with (q `as` #q)
    (select_ (#q ! #m_name `as` #manufacturer_name
          :* #q ! #p_price `as` #product_price
          :* #p ! #name `as` #product_name)
    (from (common #q 
      & leftOuterJoin 
          (table (#products `as` #p)) 
          (#p ! #manufacturer .== #q ! #m_code .&& #p ! #price .== #q ! #p_price))))
  where
    q :: Query '[] '[] DB '[]
      '[ "m_code" ::: 'NotNull 'PGint8
       , "m_name" ::: 'NotNull 'PGtext
       , "p_price" ::: 'Null 'PGmoney
       ]
    q = select_ (#m ! #code `as` #m_code
              :* #m ! #name `as` #m_name
              :* max_ (allNotNull $ #p ! #price) `as` #p_price)
        (from (table (#manufacturers `as` #m)
                   & leftOuterJoin (table (#products `as` #p)) (#p ! #manufacturer .== #m ! #code))
        & groupBy (#m ! #code :* #m ! #name))

-- | 17. Add a new product: Loudspeakers, $70, manufacturer 2.
ex17 :: Manipulation '[] DB '[] '[]
ex17 = insertInto_ #products 
        (Values_
          (Default `as` #code
        :* Set (inline "Loudspeakers") `as` #name
        :* Set (inline (Money 7000)) `as` #price
        :* Set (inline (2 :: Int64)) `as` #manufacturer)
        )

-- | 18. Update the name of product 8 to "Laser Printer".
ex18 :: Manipulation '[] DB '[] '[]
ex18 = update_ #products
        (Set (inline "Laser Printer") `as` #name)
        (#code .== 2)

-- | 19. Apply a 10% discount to all products.
ex19 :: Manipulation '[] DB '[] '[]
ex19 = update_ #products
        (Set (money `cast` ((numeric `cast` #price) * 0.9)) `as` #price)
        true

-- | 20. Apply a 10% discount to all products with a price larger than or equal to $120.
ex20 :: Manipulation '[] DB '[] '[]
ex20 = update_ #products
        (Set (money `cast` ((numeric `cast` #price) * 0.9)) `as` #price)
        (#price .>= inline (Money 120))

setup :: Migration Definition (Public '[]) DB
setup = Migration
  { name = pack "TheComputerStore"
  , migration = up
  } where
      up :: Definition (Public '[]) DB
      up = createTable #manufacturers
           (serial8 `as` #code
         :* notNullable text `as` #name)
           (primaryKey #code `as` #pk_manufacturers)
       >>> createTable #products
           (serial8 `as` #code
         :* notNullable text `as` #name
         :* notNullable money `as` #price
         :* notNullable int8 `as` #manufacturer)
           (primaryKey #code `as` #pk_products
         :* foreignKey #manufacturer #manufacturers #code OnDeleteCascade OnUpdateCascade `as` #fk_manufacturer)


exercises/EmployeeManagement.hs
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE OverloadedLabels #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE DataKinds #-}
module EmployeeManagement where

import Shared
import Squeal.PostgreSQL
import Data.Int (Int64)
import Data.Text (pack)

type instance Columns "departments" =
  '[ "code" ::: 'Def :=> 'NotNull 'PGint8
   , "name" ::: 'NoDef :=> 'NotNull 'PGtext
   , "budget" ::: 'NoDef :=> 'NotNull 'PGmoney ]

type instance Constraints "departments" =
  '[ "pk_departments" ::: 'PrimaryKey '["code"] ]

type instance Columns "employees" =
  '[ "ssn" ::: 'NoDef :=> 'NotNull 'PGint8 
   , "name" ::: 'NoDef :=> 'NotNull 'PGtext
   , "last_name" ::: 'NoDef :=> 'NotNull 'PGtext
   , "department" ::: 'NoDef :=> 'NotNull 'PGint8 ]

type instance Constraints "employees" =
  '[ "pk_employees" ::: 'PrimaryKey '["ssn"]
   , "fk_department" ::: 'ForeignKey '["department"] "departments" '["code"] ]

type DB = Public '[ SimpleTable "departments", SimpleTable "employees" ]

-- | 1. Select the last name of all employees.
ex1 :: Query '[] '[] DB '[] '[ "last_name" ::: 'NotNull 'PGtext ]
ex1 = select_ #last_name (from (table #employees))

-- | 2. Select the last name of all employees, without duplicates.
ex2 :: Query '[] '[] DB '[] '[ "last_name" ::: 'NotNull 'PGtext ]
ex2 = selectDistinct_ #last_name (from (table #employees))

-- | 3. Select all of the data of employees whose last name is "Smith".
ex3 :: Query '[] '[] DB '[]
  '[ "ssn" ::: 'NotNull 'PGint8
   , "name" ::: 'NotNull 'PGtext
   , "last_name" ::: 'NotNull 'PGtext
   , "department" ::: 'NotNull 'PGint8 ]
ex3 = select_ (#ssn :* #name :* #last_name :* #department) 
      (from (table #employees) & where_ (#name .== inline "Smith"))

-- | 4. Select all the data of employees whose last name is "Smith" or
-- "Doe"
ex4 :: Query '[] '[] DB '[]
  '[ "ssn" ::: 'NotNull 'PGint8
   , "name" ::: 'NotNull 'PGtext
   , "last_name" ::: 'NotNull 'PGtext
   , "department" ::: 'NotNull 'PGint8 ]
ex4 = select_ (#ssn :* #name :* #last_name :* #department)
      (from (table #employees) & where_ (#name .== inline "Smith" .|| #name .== inline "Doe"))

-- | 5. Select all the data of employees that work in department 14.
ex5 :: Query '[] '[] DB '[] 
  '[ "ssn" ::: 'NotNull 'PGint8
   , "name" ::: 'NotNull 'PGtext
   , "last_name" ::: 'NotNull 'PGtext
   , "department" ::: 'NotNull 'PGint8 ]
ex5 = select_ (#ssn :* #name :* #last_name :* #department)
              (from (table #employees) & where_ (#department .== inline (14 :: Int64))) 

-- | 6. Select all the data of employees that work in department 37 or department 77.
ex6 :: Query '[] '[] DB '[]
  '[ "ssn" ::: 'NotNull 'PGint8
   , "name" ::: 'NotNull 'PGtext
   , "last_name" ::: 'NotNull 'PGtext
   , "department" ::: 'NotNull 'PGint8 ]
ex6 = select_ (#ssn :* #name :* #last_name :* #department)
              (from (table #employees) & where_ (#department .== inline (37 :: Int64)))

-- | 7. Select all the data of employees whose last name begins with an "S".
ex7 :: Query '[] '[] DB '[]
  '[ "ssn" ::: 'NotNull 'PGint8
   , "name" ::: 'NotNull 'PGtext
   , "last_name" ::: 'NotNull 'PGtext
   , "department" ::: 'NotNull 'PGint8 ]
ex7 = select_ (#ssn :* #name :* #last_name :* #department)
              (from (table #employees) & where_ (#last_name `like` inline "S%"))

-- | 8. Select the sum of all the departments' budgets.
ex8 :: Query '[] '[] DB '[]
  '[ "sum_of_budgets" ::: 'Null 'PGmoney ]
ex8 = select (sum_ (All $ #d ! #budget) `as` #sum_of_budgets)
             (from (table (#departments `as` #d)) & groupBy Nil)

-- | 9. Select the number of employees in each department (you only need to
-- show the department code and the number of employees).
ex9 :: Query '[] '[] DB '[]
  '[ "number_of_employees" ::: 'NotNull 'PGint8
   , "department_code" ::: 'NotNull 'PGint8 ]
ex9 = with (q `as` #q) 
      (select_ ( count (All $ #e ! #ssn) `as` #number_of_employees
              :* #q ! #department_code `as` #department_code )
      (from (common #q
          & leftOuterJoin (table (#employees `as` #e)) (#e ! #department .== #q ! #department_code))
      & groupBy (#q ! #department_code))
      ) where
  q = select (#code `as` #department_code) (from (table #departments))

-- | 10. Select all the data of employees, including each employee's department's data.
ex10 :: Query '[] '[] DB '[]
  '[ "ssn" ::: 'NotNull 'PGint8
   , "name" ::: 'NotNull 'PGtext
   , "last_name" ::: 'NotNull 'PGtext
   , "department_code" ::: 'NotNull 'PGint8 
   , "department_name" ::: 'NotNull 'PGtext
   , "department_budget" ::: 'NotNull 'PGmoney ]
ex10 = select_ (#e ! #ssn `as` #ssn
             :* #e ! #name `as` #name
             :* #e ! #last_name `as` #last_name
             :* #e ! #department `as` #department_code
             :* #d ! #name `as` #department_name
             :* #d ! #budget `as` #department_budget)
       (from (table (#employees `as` #e)
             & innerJoin (table (#departments `as` #d)) (#d ! #code .== #e ! #department)))

-- | 11. Select the name and last name of each employee, along with the name and
-- budget of the employee's department.
ex11 :: Query '[] '[] DB '[]
  '[ "employee_name" ::: 'NotNull 'PGtext
   , "employee_last_name" ::: 'NotNull 'PGtext
   , "department_name" ::: 'NotNull 'PGtext
   , "department_budget" ::: 'NotNull 'PGmoney ]
ex11 = select_ (#e ! #name `as` #employee_name
             :* #e ! #last_name `as` #employee_last_name
             :* #d ! #name `as` #department_name
             :* #d ! #budget `as` #department_budget)
       (from (table (#employees `as` #e)
             & innerJoin (table (#departments `as` #d)) (#d ! #code .== #e ! #department)))

-- | 12. Select the name and last name of employees working for departments with a budget greater than $60,000.
ex12 :: Query '[] '[] DB '[]
  '[ "first_name" ::: 'NotNull 'PGtext
   , "last_name" ::: 'NotNull 'PGtext ]
ex12 = select_ (#e ! #name `as` #first_name :* #e ! #last_name `as` #last_name)
  (from (table (#employees `as` #e)
        & innerJoin (table (#departments `as` #d)) (#d ! #code .== #e ! #department))
  & where_ (#d ! #budget .> inline (Money 6000000)))

-- | 13. Select the departments with a budget larger than the average budget of all the departments.
ex13 :: Query '[] '[] DB '[]
  '[ "code" ::: 'NotNull 'PGint8
   , "name" ::: 'NotNull 'PGtext
   , "budget" ::: 'NotNull 'PGmoney ]
ex13 = with (q `as` #q)
  (select_ (#d ! #code `as` #code 
         :* #d ! #name `as` #name
         :* #d ! #budget `as` #budget)
    (from (common #q
          & innerJoin (table (#departments `as` #d)) true)
    & where_ (#d ! #budget .> #q ! #average_budget)))
  where
    q = select_ ((money `cast` avg (All $ money `cast` (#d ! #budget))) `as` #average_budget)
      (from (table (#departments `as` #d))
      & groupBy Nil)

-- | 14. Select the names of departments with more than two employees.
ex14 :: Query '[] '[] DB '[]
  '[ "department_name" ::: 'NotNull 'PGtext ]
ex14 = with (q `as` #q)
  (select_ (#d ! #name `as` #department_name)
    (from (common #q
          & innerJoin (table (#departments `as` #d)) (#d ! #code .== #q ! #department_code))
    & where_ (#q ! #employee_count .> 2)))
  where
    q = select_ (#d ! #code `as` #department_code
             :* count (All $ #e ! #ssn) `as` #employee_count)
             (from (table (#departments `as` #d)
                   & leftOuterJoin (table (#employees `as` #e)) (#e ! #department .== #d ! #code))
             & groupBy (#d ! #code))

-- | 15. Select the name and last name of employees working for departments with second lowest budget.
ex15 :: Query '[] '[] DB '[]
  '[ "first_name" ::: 'NotNull 'PGtext
   , "last_name" ::: 'NotNull 'PGtext ]
ex15 = with (q `as` #q)
  (select_ (#e ! #name `as` #first_name
         :* #e ! #last_name `as` #last_name)
  (from (common #q
        & innerJoin (table (#departments `as` #d)) (#d ! #code .== #q ! #department_code)
        & innerJoin (table (#employees `as` #e)) (#e ! #department .== #d ! #code))))
  where
    q = select_ (#d ! #code `as` #department_code)
        (from (table (#departments `as` #d))
        & orderBy [Asc $ #d ! #budget]
        & offset 1
        & limit 1)

-- | 16. Add a new department called "Quality Assurance", with a budget of $40,000 and departmental
-- code 11. Add an employee called "Mary Moore" in that department, with SSN 847-21-9811.
ex16 :: Manipulation '[] DB '[] '[]
ex16 = with (qa `as` #qa)
  (insertInto #employees
    (Subquery (select_ (inline (847219811 :: Int64) `as` #ssn
                     :* inline "Mary" `as` #name
                     :* inline "Moore" `as` #last_name
                     :* #qa ! #code `as` #department) (from (common #qa)))) OnConflictDoRaise (Returning_ Nil))
  where
    qa = insertInto #departments
      (Values_
        (Set (inline (11 :: Int64)) `as` #code
      :* Set (inline "Quality Assurance") `as` #name
      :* Set (inline (Money 4000000)) `as` #budget)) OnConflictDoRaise (Returning_ #code)
        

-- | 17. Reduce the budget of all departments by 10%.
ex17 :: Manipulation '[] DB '[] '[]
ex17 = update_ #departments (Set (money `cast` ((numeric `cast` #budget) * 0.9)) `as` #budget) true

-- | 18. Reassign all employees from the Research department (code 77) to the IT department (code 14).
ex18 :: Manipulation '[] DB '[] '[]
ex18 = update_ #employees (Set (inline (14 :: Int64)) `as` #department) (#department .== 77)

-- | 19. Delete from the table all employees in the IT department (code 14).
ex19 :: Manipulation '[] DB '[] '[]
ex19 = deleteFrom #employees NoUsing (#department .== 14) (Returning_ Nil)

-- | 20. Delete from the table all employees who work in departments with a
-- budget greater than or equal to $60,000.
ex20 :: Manipulation '[] DB '[] '[]
ex20 = deleteFrom #employees (Using (table (#departments `as` #d))) 
  (#d ! #budget .>= inline (Money 6000000) .&& #d ! #code .== #employees ! #department)
  (Returning_ Nil)

setup :: Migration Definition (Public '[]) DB
setup = Migration
  { name = pack "EmployeeManagement"
  , migration = up
  } where
      up :: Definition (Public '[]) DB
      up = createTable #departments
           (serial8 `as` #code
         :* notNullable text `as` #name
         :* notNullable money `as` #budget)
           (primaryKey #code `as` #pk_departments)
       >>> createTable #employees
           (notNullable int8 `as` #ssn
         :* notNullable text `as` #name
         :* notNullable text `as` #last_name
         :* notNullable int8 `as` #department)
           (primaryKey #ssn `as` #pk_employees
         :* foreignKey #department #departments #code OnDeleteCascade OnUpdateCascade `as` #fk_department)
